// @ionic/angular/dist/util/ionic-router-reuse-strategy.js
import { deepEqual, objectValues } from './util';
var IonicRouteStrategy = /** @class */ (function () {
    function IonicRouteStrategy() {
    }
    IonicRouteStrategy.prototype.shouldDetach = function (_route) {
        return false;
    };
    // tslint:disable-next-line
    IonicRouteStrategy.prototype.store = function (_route, _detachedTree) { };
    IonicRouteStrategy.prototype.shouldAttach = function (_route) {
        return false;
    };
    IonicRouteStrategy.prototype.retrieve = function (_route) {
        return null;
    };
    IonicRouteStrategy.prototype.shouldReuseRoute = function (future, curr) {
        // checking router params
        var futureParams = objectValues(future.params);
        var currParams = objectValues(curr.params);
        if (futureParams &&
            !!futureParams.length &&
            currParams &&
            currParams.length > 0) {
            // If the router params do not match, render the new component
            return (deepEqual(future.params, curr.params) &&
                future.routeConfig === curr.routeConfig);
        }
        else {
            return future.routeConfig === curr.routeConfig;
        }
    };
    return IonicRouteStrategy;
}());
export { IonicRouteStrategy };
