// @ionic/core/dist/collection/utils/theme.js
export function hostContext(selector, el) {
    return !!el.closest(selector);
}
/**
 * Create the mode and color classes for the component based on the classes passed in
 */
export function createColorClasses(color) {
    return (color) ? {
        'ion-color': true,
        [`ion-color-${color}`]: true
    } : null;
}
export function createThemedClasses(mode, name) {
    return {
        [name]: true,
        [`${name}-${mode}`]: !!mode
    };
}
export function getClassList(classes) {
    if (classes) {
        const array = Array.isArray(classes) ? classes : classes.split(' ');
        return array
            .filter(c => c != null)
            .map(c => c.trim())
            .filter(c => c !== '');
    }
    return [];
}
export function getClassMap(classes) {
    const map = {};
    getClassList(classes).forEach(c => map[c] = true);
    return map;
}
export async function openURL(win, url, ev, direction) {
    if (url && url[0] !== '#' && url.indexOf('://') === -1) {
        const router = win.document.querySelector('ion-router');
        if (router) {
            ev && ev.preventDefault();
            await router.componentOnReady();
            return router.push(url, direction);
        }
    }
    return Promise.resolve();
}
export function getParentNode(node) {
    // this also checks for host elements of shadow root node
    // if the parent node is a document fragment (shadow root)
    // then use the "host" property on it
    // otherwise use the parent node
    // DOCUMENT_FRAGMENT_NODE nodeType === 11
    const parentNode = node.parentNode;
    return (parentNode && parentNode.NODE_TYPE === 11 ? parentNode.host : parentNode);
}
